docs/madr/decisions/012-health-check-policy.md

```diff
@@ -0,0 +1,180 @@
+
+## Specification
+
+### Matching
+
+#### From level
+
+The way active health checking works is by sending outgoing requests, I don't think it makes sense to have "from" section.
+
+#### Top & to level
+
+`Top` and `to` levels look to me as mirror cases, so I think it's worth talking about them together.
+
+`MeshGatewayRoute` and `MeshHTTPRoute` do not make sense as on its own they do not have "health",
+also Envoy does set health on a route only per Cluster/Endpoint.
+
+`MeshService` is a natural fit, but I think we should be careful when allowing other targets combinations.
+
+##### Option 1 - allow everything
+
+###### Pros
+
+- Flexibility
+- Can avoid DOS-ing by using reachable services
+
+###### Cons
+
+- Users might accidentally DOS their infra by making Everything HC service X
+- Without reachable services using `targetRef=Mesh` and `to.targetRef=Mesh` does not make sense
+- Cases that might produce a lot of traffic need to be documented
+
+##### Option 2 - disallow Mesh-Mesh
+
+###### Pros
+
+- Flexibility (without 1 case)
+- Can avoid DOS-ing by using reachable services
+
+###### Cons
+
+- Users might accidentally DOS their infra by making Everything HC service X
+- Cases that might produce a lot of traffic need to be documented
+
+##### Option 3 - allow only MeshService
+
+###### Pros
+
+- Harder to accidentally DOS infrastructure by running too aggressive health checks
+
+###### Cons
+
+- Inflexibility
+
+#### Decision drivers
+
+As we don't pose any restrictions on HC right now I would go with **option 1** and an entry in docs about possible problems.
+
+### GRPC
+
+GRPC support seems pretty straight forward,
+just need to use a different health checking type - [GrpcHealthCheck](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/health_check.proto.html?highlight=grpc_health_check#envoy-v3-api-field-config-core-v3-healthcheck-grpc-health-check)
+
+### Configuration options
+
+There are some features that we currently do not expose (like pass through mode / caching, redis / thrift protocol),
+but currently this MADR does not aim to change that, if you think something critical is missing from this functionality,
+let us know in the PR.
+
+The question is - do we need `service_name`, `authority` and `initial_metadata` gRPC options in the first iteration?
+
+```yaml
+default:
+  interval: 10s
+  timeout: 2s
+  unhealthyThreshold: 3
+  healthyThreshold: 1
+  initialJitter: 5s # optional
+  intervalJitter: 6s # optional
+  intervalJitterPercent: 10 # optional
+  healthyPanicThreshold: 60 # optional, by default 50
+  failTrafficOnPanic: true # optional, by default false
+  noTrafficInterval: 10s # optional, by default 60s
+  eventLogPath: "/tmp/health-check.log" # optional
+  alwaysLogHealthCheckFailures: true # optional, by default false
+  reuseConnection: false # optional, by default true
+  tcp: # only one of tcp http or grpc can be enabled
```

Comment:
Does it make sense to not have this exclusive and just have a policy stating that it'll pick the most specific protocol depending on the service annotation?

Reply:
That makes sense, you can still disable it by override but you won't be able to change the type of HC by override - is that ok?

Reply:
You could by setting `disabled=true` on the most specific protocol no?

Reply:
> it'll pick the most specific protocol depending on the service annotation

> You could by setting disabled=true on the most specific protocol no?

So the algorithm would be:
1. service annotation = http
2. http.disabled == true
3. **try tcp** (more general proto)
4. tcp.disabled == false
5. use tcp

So it would use the annotation as a suggestion and fall back to more general protocol if needed. Yeah that would work.

Reply:
Worth adding this in the MADR

Reply:
Done 👍